pkg/integrations/v2/app_o11y_exporter/app_o11y_exporter.go

```diff
@@ -0,0 +1,206 @@
+		// Logs
+		lokiExporter,
+		// Exceptions
+		lokiExceptionExporter,
+		// Measurements
+		measurementsPromExporter,
+	}
+
+	receiver := receiver.NewAppReceiver(c.ExporterConfig, exp)
+
+	for _, e := range exp {
+		e.Init()
+	}
+
+	return &appo11yIntegration{
+		logger:          l,
+		integrationName: c.Name(),
+		instanceId:      id,
+		common:          c.Common,
+		globals:         globals,
+		conf:            c.ExporterConfig,
+		receiver:        receiver,
+		exporters:       exp,
+		reg:             reg,
+	}, nil
+}
+
+func (i *appo11yIntegration) Handler(prefix string) (http.Handler, error) {
+	r := mux.NewRouter()
+	r.Handle(path.Join(prefix, "metrics"), promhttp.HandlerFor(i.reg, promhttp.HandlerOpts{}))
+	return r, nil
+}
+
+func (ai *appo11yIntegration) Targets(ep integrations.Endpoint) []*targetgroup.Group {
+	integrationNameValue := model.LabelValue("integrations/" + ai.integrationName)
+
+	group := &targetgroup.Group{
+		Labels: model.LabelSet{
+			model.InstanceLabel: model.LabelValue(ai.instanceId),
+			model.JobLabel:      integrationNameValue,
+			"agent_hostname":    model.LabelValue(ai.globals.AgentIdentifier),
+
+			// Meta labels that can be used during SD.
+			"__meta_agent_integration_name":     model.LabelValue(ai.integrationName),
+			"__meta_agent_integration_instance": model.LabelValue(ai.instanceId),
+		},
+		Source: fmt.Sprintf("%s/%s", ai.integrationName, ai.instanceId),
+	}
+
+	group.Targets = append(group.Targets, model.LabelSet{
+		model.AddressLabel:     model.LabelValue(ep.Host),
+		model.MetricsPathLabel: model.LabelValue(path.Join(ep.Prefix, "/metrics")),
+	})
+
+	return []*targetgroup.Group{group}
+}
+
+func (ai *appo11yIntegration) ScrapeConfigs(sd discovery.Configs) []*autoscrape.ScrapeConfig {
+	if !*ai.common.Autoscrape.Enable {
+		return nil
+	}
+
+	cfg := promConfig.DefaultScrapeConfig
+	cfg.JobName = fmt.Sprintf("%s/%s", ai.integrationName, ai.instanceId)
+	cfg.Scheme = ai.globals.AgentBaseURL.Scheme
+	cfg.HTTPClientConfig = ai.globals.SubsystemOpts.ClientConfig
+	cfg.ServiceDiscoveryConfigs = sd
+	cfg.ScrapeInterval = ai.common.Autoscrape.ScrapeInterval
+	cfg.ScrapeTimeout = ai.common.Autoscrape.ScrapeTimeout
+	cfg.RelabelConfigs = ai.common.Autoscrape.RelabelConfigs
+	cfg.MetricRelabelConfigs = ai.common.Autoscrape.MetricRelabelConfigs
+
+	return []*autoscrape.ScrapeConfig{{
+		Instance: ai.common.Autoscrape.MetricsInstance,
+		Config:   cfg,
+	}}
+}
+
+func (ai *appo11yIntegration) RunIntegration(ctx context.Context) error {
+	var (
+		host = "0.0.0.0"
+		port = 3029
+	)
+	if ai.conf.Server.Host != "" {
+		host = ai.conf.Server.Host
+	}
+
+	if ai.conf.Server.Port != 0 {
+		port = ai.conf.Server.Port
+	}
+
+	r := mux.NewRouter()
+	r.Handle("/collect", ai.receiver.ReceiverHandler(&ai.logger))
+
+	ln, err := net.Listen("tcp", fmt.Sprintf("%s:%d", host, port))
+
+	if err != nil {
+		return err
+	}
+	err = http.Serve(ln, r)
+	if err != nil {
+		return err
+	}
+
+	<-ctx.Done()
```

Comment:
If `ctx` gets canceled, the integration needs to shut down. I don't think that's currently happening here.

Reply:
a6dfcf0c